### DISCUSSION TITLE
Debuging the P4-PSA-eBPF&Nikss Configurations

### DISCUSSION THREAD

**Discussion by g83wang** [NIKSS-vSwitch/nikss#104] [Q&A] [answered]
Hi Team, 

I wonder if there is a way to monitor/debug the PSA-eBPF pipeline & Nikss configuration.
Any input is appreciated.

I am trying to learn how to manage the p4-ebpf-upf routing with Nikss. 
Here is the upf implementation I am using:
https://github.com/p4lang/p4c/blob/main/backends/ebpf/psa/examples/upf.p4

What I did:
1. I deployed 4 virtual machines (A, B, C, and D) and 3 virtual subnets (netBA, netBC, and netBD). Thus, B will serve as the router in the routes from A to C and from A to D. And I configured the Linux routing tables so that I can ping C and D from A through B.
3. Then I compiled the p4-ebpf-upf implementation above and loaded it into the kernel of B as pipeline 0 successfully. 
4. After that I configure the pipeline and upf tables as follows:

```shell


/#!/bin/bash

# Attach the pipeline to the ingress port on node B from node A
# The network interface on NodeB of netAB is enp6s0

sudo nikss-ctl add-port pipe 0 dev enp6s0

# Output:
# skipping empty egress program...
# {
#     "name": "enp6s0",
#     "port_id": 3
# }

# Also, attach the pipeline to NetBC (enp7s0)

sudo nikss-ctl add-port pipe 0 dev enp7s0

# Output:
# skipping empty egress program...
# {
#     "name": "enp7s0",
#     "port_id": 4
# }

# Also to NetBD (enp8s0)

sudo nikss-ctl add-port pipe 0 dev enp8s0

# Output:
# skipping empty egress program...
# {
#     "name": "enp8s0",
#     "port_id": 5
# }

# Add table entries
# 1. table ingress_upf_process_ingress_l4port_ingress_l4port_fields
# The table will read info from hdr to meta.
# meta.upf.l4_sport = hdr.tcp/udp.sport;
# meta.upf.l4_dport = hdr.tcp/udp.dport;
# It is already setup for us. To retrieve the info: sudo nikss-ctl table get pipe 0 ingress_upf_process_ingress_l4port_ingress_l4port_fields

# 2. table ingress_upf_process_ingress_l4port_ingress_l4_src_port
# For the source port (should be a port on NodeA; port_id: 3), set src_port_range_id = 0

sudo nikss-ctl table add pipe 0 ingress_upf_process_ingress_l4port_ingress_l4_src_port action name ingress_upf_process_ingress_l4port_set_ingress_src_port_range_id key 3 data 0

# 3. table ingress_upf_process_ingress_l4port_ingress_l4_dst_port" (adding range_id 0 to port 4&5)
# set meta.upf.dst_port_range_id = 0

sudo nikss-ctl table add pipe 0 ingress_upf_process_ingress_l4port_ingress_l4_dst_port action name ingress_upf_process_ingress_l4port_set_ingress_dst_port_range_id key 4 data 0
sudo nikss-ctl table add pipe 0 ingress_upf_process_ingress_l4port_ingress_l4_dst_port action name ingress_upf_process_ingress_l4port_set_ingress_dst_port_range_id key 5 data 0

# 4. table ingress_upf_ingress_source_interface_lookup_by_port
# The table matches on istd.ingress_port (bit<unspecificed>) and set meta.upf.src=src where we provide the src (bit<4>) on RHS

sudo nikss-ctl table add pipe 0 ingress_upf_ingress_source_interface_lookup_by_port action name ingress_upf_ingress_set_source_interface key 3 data 1

# 5. table ingress_upf_ingress_session_lookup_by_teid
# Match on gtpu.teid (bit<32>) and set the meta.upf.seid which will be used on pdr_lookup
# it will not be matched because we will not use gtpu (in that way, will not enter into this control block.)
sudo nikss-ctl table add pipe 0 ingress_upf_ingress_session_lookup_by_teid action name ingress_upf_ingress_set_seid key 1 data 1

# 6. table ingress_upf_ingress_session_lookup_by_ue_ip
# The table will set different session end id for different ipv4_dstAddr. And the session end id will be used pdr_lookup
# dest == NodeC
sudo nikss-ctl table add pipe 0 ingress_upf_ingress_session_lookup_by_ue_ip action name ingress_upf_ingress_set_seid key 192.168.101.143 data 1
# dest == NodeD
sudo nikss-ctl table add pipe 0 ingress_upf_ingress_session_lookup_by_ue_ip action name ingress_upf_ingress_set_seid key 192.168.102.168 data 1

# 7. table ingress_upf_ingress_pdr_lookup
# table pdr_lookup {
#     key= {
#         meta.upf.seid:  exact; // 1
#         ipv4.srcAddr: ternary; // NodeA (192.168.100.195) 
#         ipv4.dstAddr: ternary; // NodeC (192.168.101.143) NodeD (192.168.102.168)
#         ipv4.protocol: ternary;// 1
#         meta.upf.src_port_range_id: ternary; // 0
#         meta.upf.dst_port_range_id: ternary; // 0
#         meta.upf.src : exact; // 1
#     }
#     actions = {
#         set_far_id();
#         @defaultonly drop();
#     }
#     const default_action = drop();
# }
# if the destination is NodeC, will set meta.upf.far_id to be 1

sudo nikss-ctl table add pipe 0 ingress_upf_ingress_pdr_lookup action name ingress_upf_ingress_set_far_id key 1 192.168.100.195 192.168.101.143 1 0 0 1 data 1

# if the destination is NodeD, will set meta.upf.far_id to be 2

sudo nikss-ctl table add pipe 0 ingress_upf_ingress_pdr_lookup action name ingress_upf_ingress_set_far_id key 1 192.168.100.195 192.168.102.168 1 0 0 1 data 2

# 8. table ingress_upf_ingress_far_lookup
# The table will match on the meta.upf.far_id and call far_forward(dest)
# table far_lookup {
#         key= {
#             meta.upf.far_id: exact;
#         }
#         actions = {
#             far_forward();
#             far_encap_forward();
#             drop();
#         }
#         const default_action = drop();
#     }
# if the meta.upf.far_id == 1 (the packet will go to NodeC), then it will call far_forward(1) which will set meta.upf.dest to 1

sudo nikss-ctl table add pipe 0 ingress_upf_ingress_far_lookup action name ingress_upf_ingress_far_forward key 1 data 1

# if the meta.upf.far_id == 2 (the packet will go to NodeD), then it will call far_forward(2) which will set meta.upf.dest to 2

sudo nikss-ctl table add pipe 0 ingress_upf_ingress_far_lookup action name ingress_upf_ingress_far_forward key 2 data 2

# 9. table ingress_ip_forward_ipv4_lpm
# table ipv4_lpm {
#         key= {
#               meta.upf.dest : exact;
#               meta.upf.outer_dst_addr: lpm;
#         }
#         actions = {
#             forward();
#             drop();
#         }
#         const default_action = drop();
#     }
# action forward(macAddr_t srcAddr,macAddr_t dstAddr, PortId_t port) {
# //macAddr_t srcAddr (bit 48) 
# //nodeB macaddr on netBC is 52:54:00:a3:96:b2
# //nodeB macaddr on netBD is 52:54:00:50:95:44
# //nodeC macaddr on netBC is 52:54:00:e5:a6:a1
# //nodeD macaddr on netBD is 52:54:00:aa:78:dd
# //for port, we are using 49153
#         send_to_port(ostd,port);
#         hdr.ethernet.srcAddr = srcAddr;
#         hdr.ethernet.dstAddr = dstAddr;
#         hdr.ipv4.ttl = hdr.ipv4.ttl - 1;
#     }
# if the meta.upf.dest is 1 and meta.upf.outer_dst_addr is 192.168.101.143, then it should be forwarded to NodeC
sudo nikss-ctl table add pipe 0 ingress_ip_forward_ipv4_lpm action name ingress_ip_forward_forward key 1 192.168.101.143 data 52:54:00:a3:96:b2 52:54:00:e5:a6:a1 4

#if the meta.upf.dest is 2 and meta.upf.outer_dst_addr is 192.168.102.168, then it should be forwarded to NodeD
sudo nikss-ctl table add pipe 0 ingress_ip_forward_ipv4_lpm action name ingress_ip_forward_forward key 2 192.168.102.168 data 52:54:00:50:95:44 52:54:00:aa:78:dd 5

``` 

After applying the configuration script above, I cannot ping D and C from A anymore.

TCP dump on the B's interface of netAB gives:
43 packets captured
43 packets received by filter
0 packets dropped by kernel

TCP dump on C and D gives:
0 packets captured
0 packets received by filter
0 packets dropped by kernel

So, I think the problem comes from how I configured the P4 tables by Nikss. 
Are there any obvious mistakes in my table configurations?
And Is there a way to monitor or debug this?

Thank you.

**Comment by tatry** [comment]
@g83wang 
You can compile program with `--trace` flag, then program will log what it is doing. To read these logs you can execute `bpftool prog tracelog`. But be aware that this was flag intended for testing compiler and program might be not accepted by kernel when using it. (Note: many events might be lost but at least you should see if packet is forwarded somewhere or dropped).

I don't see any obvious mistakes in table configurations, but take care of correct IP/MAC addresses when looking into packets.

For configuration example for this P4 program let's see PTF test case: https://github.com/p4lang/p4c/blob/main/backends/ebpf/tests/ptf/upf.py

Did you execute tcpdump on interfaces `netBC` and `netBD` on the B machine?

**Reply by g83wang** [reply]
Thank you very much for your reply.

I try to compile the upf.p4 again with the command below.
```shell
make --trace -f backends/ebpf/runtime/kernel.mk BPFOBJ=out.o P4FILE=upf.p4 P4C=p4c-ebpf psa
```

After I reload the program and configure the tables, here are my bpf programs:
```shell
3: cgroup_skb  tag 6deef7357e7b4530  gpl
	loaded_at 2023-06-28T11:03:38-0400  uid 0
	xlated 64B  jited 58B  memlock 4096B
4: cgroup_skb  tag 6deef7357e7b4530  gpl
	loaded_at 2023-06-28T11:03:38-0400  uid 0
	xlated 64B  jited 58B  memlock 4096B
5: cgroup_skb  tag 6deef7357e7b4530  gpl
	loaded_at 2023-06-28T11:03:38-0400  uid 0
	xlated 64B  jited 58B  memlock 4096B
6: cgroup_skb  tag 6deef7357e7b4530  gpl
	loaded_at 2023-06-28T11:03:38-0400  uid 0
	xlated 64B  jited 58B  memlock 4096B
7: cgroup_skb  tag 6deef7357e7b4530  gpl
	loaded_at 2023-06-28T11:03:43-0400  uid 0
	xlated 64B  jited 58B  memlock 4096B
8: cgroup_skb  tag 6deef7357e7b4530  gpl
	loaded_at 2023-06-28T11:03:43-0400  uid 0
	xlated 64B  jited 58B  memlock 4096B
11: cgroup_skb  tag 6deef7357e7b4530  gpl
	loaded_at 2023-06-28T11:03:45-0400  uid 0
	xlated 64B  jited 58B  memlock 4096B
12: cgroup_skb  tag 6deef7357e7b4530  gpl
	loaded_at 2023-06-28T11:03:45-0400  uid 0
	xlated 64B  jited 58B  memlock 4096B
21: sched_cls  name map_initializer  tag eb9c857a4f3cf2bb  gpl
	loaded_at 2023-06-28T11:05:33-0400  uid 0
	xlated 3192B  jited 2093B  memlock 4096B  map_ids 3,4,5,6,7,8,9,10,32
	btf_id 83
23: xdp  name xdp_func  tag 92e7a94b375fbbdf  gpl
	loaded_at 2023-06-28T11:05:33-0400  uid 0
	xlated 352B  jited 198B  memlock 4096B  map_ids 32
	btf_id 83
24: sched_cls  name tc_ingress_func  tag 90582f6ba8f755af  gpl
	loaded_at 2023-06-28T11:05:33-0400  uid 0
	xlated 15344B  jited 8999B  memlock 16384B  map_ids 31,9,30,28,29,26,27,25,8,23,6,19,5,22,24,7,18,4,17,3,32
	btf_id 83
```

When I launch the command ```sudo bpftool prog tracelog``` and ping the nodes C&D through B from A, I got nothing printed. 
I saw CONFIG_BPF_EVENTS=y in the kernel configuration, so the kernel shall work fine with logging?
Please let me know if there is something I did wrong.
And if there are some further steps I can take to investigate the issue.

> Did you execute tcpdump on interfaces netBC and netBD on the B machine?

Yes, both of them saw no packets.

Thank you.

**Reply by tatry** [reply, marked as answer]
`--trace` is a flag for p4c compiler, not for make itself. Your command should look like this (note the additional variable passed to make): `make -f backends/ebpf/runtime/kernel.mk BPFOBJ=out.o P4FILE=upf.p4 P4C=p4c-ebpf P4ARGS="--trace" psa`.

[Here](https://github.com/p4lang/p4c/blob/5892c5081b6c9499556ae2b940c62fc9f17d9699/backends/ebpf/tests/ptf/common.py#L119-L145) you can see how this makefile is used for PTF test. If program can't be loaded with `--trace` flag look for `--max-ternary-masks 3` option.

**Comment by g83wang** [comment]
Yes, the log helps me fix the routing. It works now. Thank you so much. 

Also, the VMs I deployed are Ubuntu machines and I use ```ping``` to test the routing. In that case, I cannot actually use the GTPU feature of the UPF function provided in the repo. (The ```ping``` can only work if I disable the match on ```meta.upf.outer_dst_addr: lpm;``` in ```table ipv4_lpm```. Because the field is set by ```action gtpu_decap()```. I am wondering if there is a way to use GTPU on the links in Ubuntu VM networks or how do we test the GTPU feature in a VM topology?

Are we using the module below?
https://docs.kernel.org/networking/gtp.html

Sorry, I know this question is a little out of the scope of the Nikss project. Please do not worry if we do not have an easy/quick answer at the top of our mind.  ;-)

Thank you.

**Reply by tatry** [reply]
In PTF tests for NIKSS we don't use `ping` directly. We use [Scapy](https://github.com/secdev/scapy) to build packets and test if they are correct (received packets are as expected). But our solution is rather one way testing.

I see a few possibilities to to achieve ICMP over GTP encapsulation:
1. Use NIKSS to decap packets using pass to kernel feature. See [documentation](https://github.com/p4lang/p4c/tree/main/backends/ebpf/psa#ntk-normal-packet-to-kernel) for more details.
2. Use kernel module you mentioned.
3. Use Scapy to write your own ping responder.
4. Something other?

**Reply by g83wang** [reply]
Thank you!
